openedx/core/lib/block_cache/block_structure_factory.py

```diff
@@ -0,0 +1,197 @@
+            or if the cached data is outdated for one or more of the
+            given transformers.
+        """
+
+        # Find root_block_usage_key in the cache.
+        zp_data_from_cache = cache.get(cls._encode_root_cache_key(root_block_usage_key))
+        if not zp_data_from_cache:
+            logger.debug(
+                "BlockStructure %r not found in the cache.",
+                root_block_usage_key,
+            )
+            return None
+        else:
+            logger.debug(
+                "Read BlockStructure %r from cache, size: %s",
+                root_block_usage_key,
+                len(zp_data_from_cache),
+            )
+
+        # Deserialize and construct the block structure.
+        block_relations, transformer_data, block_data_map = zunpickle(zp_data_from_cache)
+        block_structure = BlockStructureBlockData(root_block_usage_key)
+        block_structure._block_relations = block_relations
+        block_structure._transformer_data = transformer_data
+        block_structure._block_data_map = block_data_map
+
+        # Verify that the cached data for all the given transformers are
+        # for their latest versions.
+        outdated_transformers = {}
+        for transformer in transformers:
+            cached_transformer_version = block_structure._get_transformer_data_version(transformer)
+            if transformer.VERSION != cached_transformer_version:
+                outdated_transformers[transformer.name()] = "version: {}, cached: {}".format(
+                    transformer.VERSION,
+                    cached_transformer_version,
+                )
+        if outdated_transformers:
+            logger.info(
+                "Collected data for the following transformers are outdated:\n%s.",
+                '\n'.join([t_name + ": " + t_value for t_name, t_value in outdated_transformers.iteritems()]),
+            )
+            return None
+
+        return block_structure
+
+    @classmethod
+    def remove_from_cache(cls, root_block_usage_key, cache):
+        """
+        Removes the block structure for the given root_block_usage_key
+        from the given cache.
+
+        Arguments:
+            root_block_usage_key (UsageKey) - The usage_key for the root
+                of the block structure that is to be removed from
+                the given cache.
+
+            cache (django.core.cache.backends.base.BaseCache) - The
+                cache from which the block structure is to be
+                removed.
+        """
+        cache.delete(cls._encode_root_cache_key(root_block_usage_key))
+        # TODO also remove all block data?
+
+    @classmethod
+    def _encode_root_cache_key(cls, root_block_usage_key):
+        """
+        Returns the cache key to use for storing the block structure
+        for the given root_block_usage_key.
+        """
+        return "root.key." + unicode(root_block_usage_key)
```

Comment:
Nit/for the future:  Feels weird that every single method is a classmethod.  Is there a reason they aren't regular instance methods?  Then you instantiate different factories differently (maybe the cache would get passed in at instantiation time.  Maybe you could pass in an alternative `BlockStructure` class.)